### PR TITLE
Expression-based names for photos

### PR DESCRIPTION
**Pull request by alexbruy**
Add UI to configure expression-based photo names to the project properties dialog. If expression for a field is set, Input will use it to rename photos.
image

### CODE REVIEW

File: Mergin/attachment_fields_model.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                    continue
+
+                if not widget_setup.config().get("DocumentViewer", 1):
+                    continue
+
+                item_layer = QStandardItem(f"{layer.name()}")
+                item_field = QStandardItem(f"{field.name()}")
+                item_field.setData(layer_id, AttachmentFieldsModel.LAYER_ID)
+                item_field.setData(field.name(), AttachmentFieldsModel.FIELD_NAME)
+                exp, ok = QgsProject.instance().readEntry("Mergin", f"PhotoNaming/{layer_id}/{field.name()}")
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply Mergin/attachment_fields_model.py:35]: https://api.qgis.org/api/classQgsProject.html#a6ef066a702272e84a894cd494e278351

File: Mergin/project_settings_widget.py [hunk 2]
```diff
@@ -56,6 +64,11 @@ def __init__(self, parent=None):
         else:
             self.selective_sync_group.setEnabled(False)
 
+        self.model = AttachmentFieldsModel()
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply Mergin/project_settings_widget.py:67]: if you insist

File: Mergin/project_settings_widget.py [hunk 3]
```diff
@@ -56,6 +64,11 @@ def __init__(self, parent=None):
...
+        self.tree_fields.setModel(self.model)
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply Mergin/project_settings_widget.py:68]: No, if you have some schema. It is a tree view and it shows fields.

File: Mergin/utils.py [hunk 4]
```diff
@@ -775,7 +775,11 @@ def update_datasource(layer, new_path):
     """Updates layer datasource, so the layer is loaded from the new location"""
     options = QgsDataProvider.ProviderOptions()
     options.layerName = layer.name()
-    if layer.dataProvider().name() == "vectortile":
+    if layer.dataProvider().name() == "mbtilesvectortiles":
+        # For 3.31 and master
+        layer.setDataSource(f"url={new_path}&type=mbtiles", layer.name(), layer.dataProvider().name(), options)
+    elif layer.dataProvider().name() == "vectortile":
+        # For 3.22
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply Mergin/utils.py:782]: Fixing logic for different QGIS versions :-)

**Review by tomasMizera — APPROVED** [review]
Looking good from the code perspective

File: Mergin/attachment_fields_model.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            if layer.type() != QgsMapLayer.VectorLayer:
+                continue
+
+            for field in layer.fields():
+                widget_setup = field.editorWidgetSetup()
+                if widget_setup.type() != "ExternalResource":
+                    continue
+
+                if not widget_setup.config().get("DocumentViewer", 1):
+                    continue
```
Comment (tomasMizera) [inline Mergin/attachment_fields_model.py:29]: We need to remove the check for `DocumentViewer` here as well as we removed it in mobile

**Pull request merged**